### PR TITLE
Exponential Bucket Histogram - part 1

### PR DESCRIPTION
**Pull request by reyang**
This PR covers the implementation of [base-2 exponential histogram](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exponential-bucket-histogram-aggregation) bucket lookup algorithm based on IEEE 754 float64 optimization.

I plan to cover exponential bucket histogram in a series of PRs:
1. IEEE 754 float 64 bucket lookup algorithm and tests (focusing on .NET6 for now, everything will be internal only - no public API surface change) #3462
2. Automatic dynamic range adjustment and the ability to merge exponential histograms
3. Console exporter support
4. In-memory exporter support
5. Plumbing work with existing OpenTelemetry SDKs (e.g. configuration)
6. Additional optimizations and corner case handling (e.g. lookup table which might require finding a perfect hash, the TODO items in step 1)
7. Benchmarks and stress tests, potentially port the code to runtimes < net6.0 if there is a need

Future work that I haven't planned for at this moment:
1. Integration with Prometheus Exporter
2. Integration with OTLP Exporter

### CODE REVIEW

**Comment by alanwest** [comment]
Very much not to do with this PR, but this test has failed two times in a row... looking more closely at it

https://github.com/open-telemetry/opentelemetry-dotnet/runs/7433955618
```
Failed OpenTelemetry.Exporter.ZPages.Tests.ZPagesExporterTests.CheckingZPagesProcessor [58 ms]
  Error Message:
   System.Net.Http.HttpRequestException : Connection refused (localhost:7284)
---- System.Net.Sockets.SocketException : Connection refused
  Stack Trace:
     at System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(String host, Int32 port, HttpRequestMessage initialRequest, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.AddHttp11ConnectionAsync(HttpRequestMessage request)
   at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at OpenTelemetry.Exporter.ZPages.Tests.ZPagesExporterTests.CheckingZPagesProcessor() in /home/runner/work/opentelemetry-dotnet/opentelemetry-dotnet/test/OpenTelemetry.Exporter.ZPages.Tests/ZPagesExporterTests.cs:line 1[67](https://github.com/open-telemetry/opentelemetry-dotnet/runs/7433955618?check_suite_focus=true#step:5:68)
--- End of stack trace from previous location ---
----- Inner Stack Trace -----
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|277_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(String host, Int32 port, HttpRequestMessage initialRequest, Boolean async, CancellationToken cancellationToken)
```

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry/Metrics/ExponentialBucketHistogram.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+namespace OpenTelemetry.Metrics
+{
+    /// <summary>
+    /// Represents an exponential bucket histogram with base = 2 ^ (2 ^ (-scale)).
+    /// An exponential bucket histogram has infinite number of buckets, which are
+    /// identified by <c>Bucket[i] = ( base ^ i, base ^ (i + 1) ]</c>, where <c>i</c>
+    /// is an integer.
+    /// </summary>
+    internal class ExponentialBucketHistogram
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/ExponentialBucketHistogram.cs:32]: nit: Can we seal this? Some slight perf benefit to do that.
